fsspec/implementations/reference.py

```diff
@@ -190,68 +147,31 @@ def _key_to_record(self, key):
 
     def _get_chunk_sizes(self, field):
         if field not in self.chunk_sizes:
-            zarray = self._get_and_cache_metadata(f"{field}/.zarray")
+            zarray = self.zmetadata[f"{field}/.zarray"]
             size_ratio = self.np.array(zarray["shape"]) / self.np.array(
                 zarray["chunks"]
             )
             self.chunk_sizes[field] = self.np.ceil(size_ratio).astype(int)
         return self.chunk_sizes[field]
 
-    def _generate_record(self, field, irec):
-        refs = self.open_refs(self.join(self.root, field, f"refs.{irec}.parq"))
-        refs_type = ""
-        if "path" in refs:
-            refs_type += "url"
-            paths = refs["path"]
-            offsets = refs["offset"]
-            sizes = refs["size"]
-            if hasattr(paths, "codes"):
-                # Since this is meant to be temporarily iterated over we don't
-                # need to mind fully expanded categoricals
-                paths = self.np.asarray(paths)
-        if "raw" in refs:
-            refs_type += "raw"
-            raws = refs["raw"]
-        if refs_type == "url":
-            # Only urls
-            for i in range(paths.size):
-                yield [paths[i], offsets[i], sizes[i]]
-        elif refs_type == "raw":
-            # Only raws
-            for i in range(raws.size):
-                yield raws[i]
-        else:
-            # Mix of raw and urls
-            for i in range(raws.size):
-                raw = raws[i]
-                if raw:
-                    yield raw
-                else:
-                    yield [paths[i], offsets[i], sizes[i]]
-
-    def _generate_all_records(self, field):
-        chunk_size = self._get_chunk_sizes(field)
-        nrec = int(self.np.ceil(self.np.product(chunk_size) / self.record_size))
-        for irec in range(nrec):
-            record = self._generate_record(field, irec)
-            yield from record
-
-    def values(self):
-        return RefsValuesView(self)
-
-    def items(self):
-        return RefsItemsView(self)
-
     def __getitem__(self, key):
-        if key in self._items and key != ".zmetadata":
+        if key in self._items:
             return self._items[key]
+        if key in self.zmetadata:
+            return self.zmetadata[key]
         return self._load_one_key(key)
 
     def __setitem__(self, key, value):
         self._items[key] = value
 
     def __delitem__(self, key):
-        del self._items[key]
+        if key in self._items:
+            del self._items[key]
+        elif key in self.zmetadata:
+            del self.zmetadata[key]
+        else:
+            # set appropriate reference to zeros in memory
+            raise KeyError
```

Comment:
I didn't give this part too much thought when implementing it, but I am wondering how this would work, mainly because it's possible for references to be purged from memory anyway through the LRU cache, so if a delete operation set them to null, they could end up being reset to their original values again later unknowingly which would make this op inconsistent. Unless you know of a use-case where it is necessary to be able to delete or null refs lazily, I think just defaulting to a KeyError here is fine. Could also set `self._items[key]` to `None`?

Reply:
You are definitely right that deleting real non-metadata references seems unlikely, so we could defer this. However, it would be nice to have editable reference sets, as the current implementation is, so that you can in theory change out parts of your data ( http://martindurant.github.io/blog/mutable-kerchunk/ ). In that case, we need to keep track of all amended/deleted references, and indeed, I would I suppose use self._items to do that. Since the current lookup looks there first, it will have priority over the original, and we just have to check for None (which would raise KeyError on access).

Reply:
Thanks for sharing that, now I see what you are getting at. So for now we could do:

- Set `self._item[key] = None`
- Raise KeyError in `__getitem__`  if  `self._item[key] is None`

Then I suppose later follow-up work would be to implement a method to RefFS that writes the updated references to the parquet files?

Reply:
Exactly. We should even know exactly which files need rewriting, the rest to be left in place or copied if the new version is elsewhere. This will be great for appending.